torch/csrc/distributed/rpc/init.cpp

```diff
@@ -290,6 +292,54 @@ If the future completes with an error, an exception is thrown.
       py::arg("tensors"),
       py::arg("rf") = nullptr);
 
+  // TODO This python future wrapper wraps c10::ivalue::Future.
+  // Will merge with JIT PythonFutureWrapper while merging generic Future with
+  // c10::ivalue::Future later on.
+  struct PythonFutureWrapper {
+    explicit PythonFutureWrapper(c10::intrusive_ptr<c10::ivalue::Future> fut)
+        : fut(std::move(fut)) {}
+
+    c10::intrusive_ptr<c10::ivalue::Future> fut;
+  };
+
+  // Since FutureMessage is binded to Future, here we need to bind the
+  // PythonFutureWrapper to a different name.
+  // TODO Once python object can be tagged as IValue and c10::ivalue::Future is
+  // implemented as generic Future<IValue>, we can consider all rpc call
+  // to return a future<IValue> later on.
+  shared_ptr_class_<PythonFutureWrapper>(module, "_pyFuture")
+      .def("wait", [](PythonFutureWrapper& fut) {
+        fut.fut->wait();
+        auto res = fut.fut->value();
+        {
+          // acquiring GIL as torch::jit::toPyObject creates new py::object
+          // without grabbing the GIL.
+          AutoGIL ag;
```

Comment:
By default, pybind uses acquires GIL when calling C++ functions.

See
https://pybind11.readthedocs.io/en/stable/advanced/misc.html#global-interpreter-lock-gil

So you may want to add a tag, `py::gil_scoped_release`, to increase efficiency of this member function.

Reply:
that is a good point!

Reply:
right, looks like it is re-entrant for the same thread. But it could cause deadlock if another thread is trying to acquire the lock and this thread is waiting for another thread to finish. (just got it from debugging the flaky tests:(). So I think it is always a good practice to tag gil release if there are some python stuff called in C++